Repeated failure to create a container can trigger a panic

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**

It is possible for buildah to panic due a division by zero error if it is unable to create new containers via `store.CreateContainer`.

**Steps to reproduce the issue:**

It's not that clear, but here's a rough description of what we're doing - we're invoking `buildah` via `podman` as part of a `bazel` build. That build will likely be launching multiple podman processes. At some point, one of the `build` commands fails with:

`panic: runtime error: integer divide by zero`

It looks like this is possible if `store.CreateContainer` continually fails for a reason that doesn't match the `!errors.Is(err, storage.ErrDuplicateName) || options.Container != ""` predicate. If that happens frequently enough, `conflict` wraps to `0`, and the `%` operator here `tmpName = fmt.Sprintf("%s-%d", name, rand.Int()%conflict)` panics. This is awkward, because it looks like we discard the error at that point, preventing further debugging.

Interestingly, if I restrict bazel to a single job (i.e single podman instance) the problem goes away, so it seems that is a contributing factor.

If it helps, the wrapper we've got invoking podman is doing something like this:

```
podman --log-level=info build \
  --rm --quiet\
  --build-arg "version=${version}" \
  --label "build_number=${build_number}" \
  --label "revision=${revision}" \
  --build-arg "build_number=${build_number}" \
  --build-arg "revision=${revision}" \
  --label "built-with-tag=${image_tag}" \
  --tag="${image_tag}" \
  --file="${containerfile}" \
  $NON_SYMLINK_DIR
```

with appropriate variables.


**Describe the results you received:**

```
panic: runtime error: integer divide by zero

goroutine 100 [running]:
panic({0x564cd87b9a20, 0x564cd9408e90})
	/usr/lib/golang/src/runtime/panic.go:941 +0x397 fp=0xc0018922c0 sp=0xc001892200 pc=0x564cd719b777
runtime.panicdivide()
	/usr/lib/golang/src/runtime/panic.go:199 +0x45 fp=0xc0018922e0 sp=0xc0018922c0 pc=0x564cd719a0c5
github.com/containers/podman/vendor/github.com/containers/buildah.newBuilder({_, _}, {_, _}, {0xc00041eed0, {0xc0013bc5c0, 0x3c}, {0x0, 0x0}, {0x0, ...}, ...})
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/new.go:268 +0x28f9 fp=0xc001892b18 sp=0xc0018922e0 pc=0x564cd7d95cb9
github.com/containers/podman/vendor/github.com/containers/buildah.NewBuilder({_, _}, {_, _}, {0xc00041eed0, {0xc0013bc5c0, 0x3c}, {0x0, 0x0}, {0x0, ...}, ...})
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/buildah.go:383 +0xdb fp=0xc001892ce8 sp=0xc001892b18 pc=0x564cd7d7cc1b
github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah.(*StageExecutor).prepare(0xc0013be000, {0x564cd89b2d08, 0xc000c37da0}, {0xc0013bc5c0?, 0xc00133e050?}, 0x1, 0x1, 0x0)
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah/stage_executor.go:756 +0x7c5 fp=0xc001893850 sp=0xc001892ce8 pc=0x564cd7e26125
github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah.(*StageExecutor).Execute(0xc0013be000, {0x564cd89b2d08, 0xc000c37da0}, {0xc0013bc5c0, 0x3c})
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah/stage_executor.go:946 +0x6a9 fp=0xc001893da0 sp=0xc001893850 pc=0x564cd7e27d89
github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah.(*Executor).buildStage(0xc000452480, {0x564cd89b2d08, 0xc000c37da0}, 0x1?, {0xc00041ef30, 0x1, 0x1}, 0x0)
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah/executor.go:504 +0x432 fp=0xc001893e78 sp=0xc001893da0 pc=0x564cd7e1b992
github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah.(*Executor).Build.func3.1()
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah/executor.go:809 +0x2c5 fp=0xc001893fe0 sp=0xc001893e78 pc=0x564cd7e1eec5
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc001893fe8 sp=0xc001893fe0 pc=0x564cd71cf981
created by github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah.(*Executor).Build.func3
	/builddir/build/BUILD/containers-podman-bfac4a5/_build/src/github.com/containers/podman/vendor/github.com/containers/buildah/imagebuildah/executor.go:780 +0x34a
```

**Describe the results you expected:**

It would have been lovely to have seen the error that caused this rather than the panic. If that's logged somewhere else (a quick look at the source suggested to me that it wasn't), or there's a different place I could look for this, please let me know.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
$ rpm -q buildah
package buildah is not installed
```

I think that is expected as we're coming from podman, so I'm going to skip to the podman section

**Output of `podman version` if reporting a `podman build` issue:**

```
$ podman --version
podman version 4.2.0
```

**Output of `cat /etc/*release`:**

```
$ cat /etc/redhat-release 
Rocky Linux release 8.7 (Green Obsidian)

```

**Output of `uname -a`:**

```
$ uname -a
Linux 4.18.0-425.10.1.el8_7.x86_64 #1 SMP Thu Jan 12 16:32:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library. The storage.conf file
# overrides all other storage.conf files. Container engines using the
# container/storage library do not inherit fields from other storage.conf
# files.
#
#  Note: The storage.conf file overrides other storage.conf files based on this precedence:
#      /usr/containers/storage.conf
#      /etc/containers/storage.conf
#      $HOME/.config/containers/storage.conf
#      $XDG_CONFIG_HOME/containers/storage.conf (If XDG_CONFIG_HOME is set)
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver, Must be set for proper operation.
driver = "overlay"

# Temporary storage location
runroot = "/run/containers/storage"

# Primary Read/Write location of container storage
# When changing the graphroot location on an SELINUX system, you must
# ensure  the labeling matches the default locations labels with the
# following commands:
# semanage fcontext -a -e /var/lib/containers/storage /NEWSTORAGEPATH
# restorecon -R -v /NEWSTORAGEPATH
graphroot = "/var/lib/containers/storage"


# Storage path for rootless users
#
# rootless_storage_path = "$HOME/.local/share/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Allows specification of how storage is populated when pulling images. This
# option can speed the pulling process of images compressed with format
# zstd:chunked. Containers/storage looks for files within images that are being
# pulled from a container registry that were previously pulled to the host.  It
# can copy or create a hard link to the existing file when it finds them,
# eliminating the need to pull them from the container registry. These options
# can deduplicate pulling of content, disk storage of content and can allow the
# kernel to use less memory when running containers.

# containers/storage supports four keys
#   * enable_partial_images="true" | "false"
#     Tells containers/storage to look for files previously pulled in storage
#     rather then always pulling them from the container registry.
#   * use_hard_links = "false" | "true"
#     Tells containers/storage to use hard links rather then create new files in
#     the image, if an identical file already existed in storage.
#   * ostree_repos = ""
#     Tells containers/storage where an ostree repository exists that might have
#     previously pulled content which can be used when attempting to avoid
#     pulling content from the container registry
pull_options = {enable_partial_images = "false", use_hard_links = "false", ostree_repos=""}

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to the UIDs/GIDs as they should appear outside of the container,
# and the length of the range of UIDs/GIDs.  Additional mapped sets can be
# listed and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and then a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped in-container ID,
# until all of the entries have been used for maps.
#
# remap-user = "containers"
# remap-group = "containers"

# Root-auto-userns-user is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid and /etc/subgid file.  These ranges will be partitioned
# to containers configured to create automatically a user namespace.  Containers
# configured to automatically create a user namespace can still overlap with containers
# having an explicit mapping set.
# This setting is ignored when running as rootless.
# root-auto-userns-user = "storage"
#
# Auto-userns-min-size is the minimum size for a user namespace created automatically.
# auto-userns-min-size=1024
#
# Auto-userns-max-size is the minimum size for a user namespace created automatically.
# auto-userns-max-size=65536

[storage.options.overlay]
# ignore_chown_errors can be set to allow a non privileged user running with
# a single UID within a user namespace to run containers. The user can pull
# and use any image even those with multiple uids.  Note multiple UIDs will be
# squashed down to the default uid in the container.  These images will have no
# separation between the users in the container. Only supported for the overlay
# and vfs drivers.
#ignore_chown_errors = "false"

# Inodes is used to set a maximum inodes of the container image.
# inodes = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev,metacopy=on"

# Set to skip a PRIVATE bind mount on the storage home directory.
# skip_mount_home = "false"

# Size is used to set a maximum size of the container image.
# size = ""

# ForceMask specifies the permissions mask that is used for new files and
# directories.
#
# The values "shared" and "private" are accepted.
# Octal permission masks are also accepted.
#
#  "": No value specified.
#     All files/directories, get set with the permissions identified within the
#     image.
#  "private": it is equivalent to 0700.
#     All files/directories get set with 0700 permissions.  The owner has rwx
#     access to the files. No other users on the system can access the files.
#     This setting could be used with networked based homedirs.
#  "shared": it is equivalent to 0755.
#     The owner has rwx access to the files and everyone else can read, access
#     and execute them. This setting is useful for sharing containers storage
#     with other users.  For instance have a storage owned by root but shared
#     to rootless users as an additional store.
#     NOTE:  All files within the image are made readable and executable by any
#     user on the system. Even /etc/shadow within your image is now readable by
#     any user.
#
#   OCTAL: Users can experiment with other OCTAL Permissions.
#
#  Note: The force_mask Flag is an experimental feature, it could change in the
#  future.  When "force_mask" is set the original permission mask is stored in
#  the "user.containers.override_stat" xattr and the "mount_program" option must
#  be specified. Mount programs like "/usr/bin/fuse-overlayfs" present the
#  extended attribute permissions to processes within containers rather then the
#  "force_mask"  permissions.
#
# force_mask = ""

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# metadata_size is used to set the `pvcreate --metadatasize` options when
# creating thin devices. Default is 128k
# metadata_size = ""

# Size is used to set a maximum size of the container image.
# size = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"
```

--- a/new.go
+++ b/new.go
@@ -239,7 +239,7 @@ func newBuilder(ctx context.Context, store storage.Store, options BuilderOptions
 		tmpName = findUnusedContainer(tmpName, containers)
 	}
 
-	conflict := 100
+	suffixDigitsModulo := 100
 	for {
 
 		var flags map[string]interface{}
@@ -265,8 +265,8 @@ func newBuilder(ctx context.Context, store storage.Store, options BuilderOptions
 		if !errors.Is(err, storage.ErrDuplicateName) || options.Container != "" {
 			return nil, fmt.Errorf("creating container: %w", err)
 		}
-		tmpName = fmt.Sprintf("%s-%d", name, rand.Int()%conflict)
-		conflict = conflict * 10
+		tmpName = fmt.Sprintf("%s-%d", name, rand.Int()%suffixDigitsModulo)
+		suffixDigitsModulo = suffixDigitsModulo * 10
 	}
 	defer func() {
 		if err != nil {

--- a/new.go
+++ b/new.go
@@ -266,7 +266,9 @@ func newBuilder(ctx context.Context, store storage.Store, options BuilderOptions
 			return nil, fmt.Errorf("creating container: %w", err)
 		}
 		tmpName = fmt.Sprintf("%s-%d", name, rand.Int()%suffixDigitsModulo)
-		suffixDigitsModulo = suffixDigitsModulo * 10
+		if suffixDigitsModulo < 1_000_000_000 {
+			suffixDigitsModulo *= 10
+		}
 	}
 	defer func() {
 		if err != nil {

